specifications/LearnProofs/FindHighest.tla

```diff
@@ -0,0 +1,243 @@
+\* prove something that isn't true!
+TypeOK ==
+  /\ f \in Seq(Nat)
+  /\ i \in 1..(Len(f) + 1)
+  /\ i \in Nat
+  /\ h \in Nat \cup {-1}
+
+\* It's useful to prove the type invariant first, so it can be used as an
+\* assumption in further proofs to restrict variable values.
+THEOREM TypeInvariantHolds == Spec => []TypeOK
+\* To prove theorems like Spec => []Invariant, you have to:
+\*  1. Prove Invariant holds in the initial state (usually trivial)
+\*  2. Prove Invariant holds when variables are unchanged (usually trivial)
+\*  3. Prove that assuming Invariant is true, a Next step implies Invariant'
+\* The last one (inductive case) is usually quite difficult. It helps to
+\* never forget you have an extremely powerful assumption: that Invariant is
+\* true!
+PROOF
+  \* The base case
+  <1>a. Init => TypeOK
+    BY DEFS Init, TypeOK
+  \* The stuttering case
+  <1>b. TypeOK /\ UNCHANGED vars => TypeOK'
+    BY DEFS TypeOK, vars
+  \* The inductive case; usually requires breaking down Next into disjuncts
+  <1>c. TypeOK /\ Next => TypeOK'
+    <2>a. TypeOK /\ lb => TypeOK'
+      \* SUFFICES steps transform P => Q goals into Q goals, while assuming P
+      <3> SUFFICES  ASSUME TypeOK, lb
+                    PROVE TypeOK'
+          OBVIOUS
+      \* Making this a non-named step expands TypeOK and lb definitions for
+      \* all subsequent & recursive steps in this sub-proof, without requiring
+      \* the step to explicitly use BY DEFS TypeOK, lb
+      <3> USE DEFS TypeOK, lb
+      <3>a. CASE i <= Len(f) BY DEF max
+      <3>b. CASE ~(i <= Len(f))
+        <4>a. UNCHANGED <<f, h, i>> BY <3>b
+        <4> QED BY <3>b, <4>a DEF lb
+      <3> QED BY <3>a, <3>b
+    <2>b. TypeOK /\ Terminating => TypeOK'
+      BY DEFS TypeOK, Terminating, vars
+    <2> QED BY <2>a, <2>b DEF Next
+  <1> QED BY PTL, <1>a, <1>b, <1>c DEF Spec
+
+\* The inductive invariant; writing these is an art. You want an invariant
+\* that can be shown to be true in every state, and if it's true in all
+\* states, it can be shown to imply algorithm correctness as a whole.
+InductiveInvariant ==
+  \A idx \in 1..(i - 1) : f[idx] <= h
+
+THEOREM InductiveInvariantHolds == Spec => []InductiveInvariant
+PROOF
+  <1>a. Init => InductiveInvariant
+    BY DEFS Init, InductiveInvariant
+  <1>b. InductiveInvariant /\ UNCHANGED vars => InductiveInvariant'
+    BY DEFS InductiveInvariant, vars
+  \* Here we introduce TypeOK and TypeOK' as powerful assumptions, since we
+  \* have already proved that Spec => []TypeOK. Note that TypeOK is a
+  \* separate assumption from TypeOK' - including both can be the difference
+  \* between a proof being impossible or trivial!
+  <1>c. InductiveInvariant /\ TypeOK /\ TypeOK' /\ Next => InductiveInvariant'
+    <2>a. InductiveInvariant /\ Terminating => InductiveInvariant'
+      BY DEFS InductiveInvariant, Terminating, vars
+    <2>b. InductiveInvariant /\ TypeOK /\ TypeOK' /\ lb => InductiveInvariant'
```

Comment:
Somehow the inclusion of TypeOK' appears to confuse the provers. The following works:
     <2>b. InductiveInvariant /\ TypeOK /\ lb => InductiveInvariant'
       BY DEFS InductiveInvariant, TypeOK, lb, max